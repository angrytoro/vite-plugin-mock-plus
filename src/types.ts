
/** 发送 SSE 消息的函数类型 */
export type SseSender = (event: string, data: string | object) => void;

/** SSE 配置项 */
export interface MockSseStreamOptions {
  /**
   * 事件流生成器。插件会传入一个 `send` 函数，你需要调用它来发送事件。
   * 你需要自己管理何时关闭流（例如通过 clearInterval）。
   * @param send - 用于发送消息的函数 `(eventName, data) => void`
   */
  generator: (send: SseSender, close: () => void) => void;
}

/** HTTP Mock 配置项 */
export interface MockHttpItem {
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'OPTIONS';
  /**
   * 响应函数，可以根据请求动态返回结果。
   * @param req - Node.js 的 IncomingMessage 对象
   */
  response: Record<string, any> | ((req: {
    query: Record<string, any>;
    body: Record<string, any>;
    params: Record<string, string>;
  }) => any);
  statusCode?: number;
  delay?: number; // 毫秒
  headers?: Record<string, string>;
}

/** SSE Mock 配置项 */
export interface MockSseItem {
  method: 'SSE';
  stream: MockSseStreamOptions;
}

/** 单个 Mock 配置，可以是 HTTP 或 SSE */
export type MockItem = MockHttpItem | MockSseItem;

/** Mock 配置文件导出的格式 */
export type MockConfig = Record<string, MockItem>;

/** 插件配置选项 */
export interface PluginOptions {
  /**
   * Mock 文件所在的目录
   * @default 'mock'
   */
  mockDir?: string;
  /**
   * 是否在控制台打印日志
   * @default true
   */
  logger?: boolean;
  /**
   * mock 文件后缀名，默认为 '.mock'
   * @default '.mock'
   */
  fileSuffix?: string;
}
