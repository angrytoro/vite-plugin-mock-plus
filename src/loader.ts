// src/loader.ts
import fg from 'fast-glob';
import fs from 'fs';
import type { MockConfig, MockItem } from './types.js';

// 使用 Map 存储，方便快速查找和更新
export const mockStore = new Map<string, MockItem>();

export async function loadMocks(mockDir: string, fileSuffix: string = '.mock'): Promise<void> {
  mockStore.clear();
  const mockFiles = await fg(`${mockDir}/**/*${fileSuffix}.{js,json}`, {
    ignore: ['**/node_modules/**'],
    absolute: true,
  });

  for (const file of mockFiles) {
    let config: MockConfig;
    if (file.endsWith('.json')) {
      config = JSON.parse(fs.readFileSync(file, 'utf-8'));
    } else {
      // 通过在路径后添加时间戳来绕过 Node.js 的 import 缓存
      const module = await import(`${file}?t=${Date.now()}`);
      config = module.default;
    }

    for (const url in config) {
      mockStore.set(url, config[url]);
    }
  }
}
