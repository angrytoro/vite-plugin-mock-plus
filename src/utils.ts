// src/utils.ts
import type { ServerResponse } from 'http';

export const log = (message: string) => console.log(`[vite-plugin-mock-plus] ${message}`);

/**
 * 格式化并发送 SSE 消息
 * @param res - ServerResponse 对象
 * @param event - 事件名称
 * @param data - 要发送的数据
 */
export const sendSseMessage: (res: ServerResponse, event: string, data: string | object) => void = (res, event, data) => {
  const message = typeof data === 'string' ? data : JSON.stringify(data);
  res.write(`event: ${event}\n`);
  res.write(`data: ${message}\n\n`);
};
