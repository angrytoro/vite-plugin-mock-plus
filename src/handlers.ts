// src/handlers.ts
import type { IncomingMessage, ServerResponse } from 'http';
import type { MockHttpItem, MockSseItem } from './types';
import { sendSseMessage } from './utils';
import { URLSearchParams } from 'url';

// 辅助函数：解析请求体
async function parseBody(req: IncomingMessage): Promise<Record<string, any>> {
  return new Promise((resolve) => {
    let body = '';
    req.on('data', (chunk) => (body += chunk));
    req.on('end', () => {
      try {
        resolve(body ? JSON.parse(body) : {});
      } catch (e) {
        resolve({});
      }
    });
  });
}

// 处理 HTTP 请求
export async function handleHttpRequest(
  item: MockHttpItem,
  req: IncomingMessage,
  res: ServerResponse
) {
  const { response, statusCode = 200, delay = 0, headers } = item;

  const body = await parseBody(req);
  const query = Object.fromEntries(new URLSearchParams(req.url?.split('?')[1] || ''));
  
  const responseData = response({ query, body, params: {} }); // params can be added with more complex routing

  setTimeout(() => {
    res.statusCode = statusCode;
    if (headers) {
      for (const key in headers) {
        res.setHeader(key, headers[key]);
      }
    }
    res.setHeader('Content-Type', 'application/json');
    res.end(JSON.stringify(responseData));
  }, delay);
}

// 处理 SSE 请求
export function handleSseRequest(
  item: MockSseItem,
  req: IncomingMessage,
  res: ServerResponse
) {
  const { stream } = item;

  res.writeHead(200, {
    'Content-Type': 'text/event-stream',
    'Cache-Control': 'no-cache',
    Connection: 'keep-alive',
  });

  const send = (event: string, data: any) => {
    sendSseMessage(res, event, data);
  };
  
  stream.generator(send, () => {
    res.end();
  });

  // 当客户端关闭连接时，清理资源
  req.on('close', () => {
    res.end();
  });
}
