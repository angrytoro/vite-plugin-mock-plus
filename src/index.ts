// src/index.ts
import type { Plugin, ViteDevServer } from 'vite';
import type { PluginOptions } from './types';
import { mockStore, loadMocks } from './loader';
import { handleHttpRequest, handleSseRequest } from './handlers';
import { log } from './utils';
import path from 'path';
import type { IncomingMessage, ServerResponse } from 'http';

export function viteMockPro(options: PluginOptions = {}): Plugin {
  const { mockDir = 'mock', logger = true, fileSuffix = '.mock' } = options;
  const resolvedMockDir = path.resolve(process.cwd(), mockDir);

  return {
    name: 'vite-plugin-mock-plus',
    
    // 核心：注入中间件到 Vite 开发服务器
    async configureServer(server: ViteDevServer) {
      // 1. 启动时加载 mocks
      await loadMocks(resolvedMockDir, fileSuffix);
      if (logger) {
        log('Mock files loaded.');
      }

      // 2. 设置 HMR (热更新)
      server.watcher.add(resolvedMockDir); // resolvedMockDir 已为 string
      server.watcher.on('all', async (_event, file) => {
        if (file && file.startsWith(resolvedMockDir)) {
          if (logger) {
            log(`Mock file changed: ${path.basename(file)}, reloading...`);
          }
          await loadMocks(resolvedMockDir, fileSuffix);
          if (logger) {
            log('Mocks reloaded successfully.');
          }
        }
      });
      
      // 3. 中间件逻辑
      const middleware = (
        req: IncomingMessage & { url?: string; method?: string },
        res: ServerResponse,
        next: () => void
      ) => {
        const url = req.url?.split('?')[0];
        if (!url) return next();

        const mockItem = mockStore.get(url);
        if (!mockItem) return next();

        const reqMethod = req.method?.toUpperCase();

        if (mockItem.method === 'SSE' && reqMethod === 'GET') {
          if (logger) log(`[SSE] Matched: ${url}`);
          handleSseRequest(mockItem, req, res);
        } else if (
          mockItem.method !== 'SSE' &&
          (mockItem.method === reqMethod || !mockItem.method) // 如果没指定 method，则匹配所有 HTTP 方法
        ) {
          if (logger) log(`[HTTP] Matched: ${reqMethod} ${url}`);
          handleHttpRequest(mockItem, req, res);
        } else {
          // 方法不匹配
          return next();
        }
      };

      server.middlewares.use(middleware);
    },
  };
}

export default viteMockPro;
