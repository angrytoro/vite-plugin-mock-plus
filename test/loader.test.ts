// tests/loader.test.ts

import { describe, it, expect, beforeEach } from 'vitest';
import path from 'path';
import { loadMocks, mockStore } from '../src/loader'; // 引入要测试的函数和存储实例

describe('Mock Loader', () => {

  // 在每个测试用例运行之前，确保 mockStore 是干净的，
  // 这可以防止测试用例之间产生相互影响。
  beforeEach(() => {
    mockStore.clear();
  });

  it('should load mocks from a single file correctly', async () => {
    // 1. 验证初始状态
    expect(mockStore.size).toBe(0);

    // 2. 执行加载操作
    const singleFileDir = path.resolve(__dirname, 'mocks/loader-single');
    await loadMocks(singleFileDir);

    // 3. 断言结果
    expect(mockStore.size).toBe(2); // 确认加载了两个 mock 配置
    expect(mockStore.has('/api/test/user')).toBe(true);
    expect(mockStore.has('/api/test/settings')).toBe(true);

    const userMock = mockStore.get('/api/test/user');
    expect(userMock?.method).toBe('GET');
  });

  it('should load and merge mocks from multiple files', async () => {
    // 1. 验证初始状态
    expect(mockStore.size).toBe(0);

    // 2. 执行加载操作
    const multipleFilesDir = path.resolve(__dirname, 'mocks/loader-multiple');
    await loadMocks(multipleFilesDir);

    // 3. 断言结果
    expect(mockStore.size).toBe(2); // 确认合并了两个文件的配置
    expect(mockStore.has('/api/test/profile')).toBe(true); // 来自 profile.mock.ts
    expect(mockStore.has('/api/test/posts')).toBe(true);   // 来自 posts.mock.ts
  });

  it('should clear old mocks before loading new ones (HMR simulation)', async () => {
    // 场景模拟：
    // 1. 第一次加载，载入一组 mocks
    // 2. 第二次加载（模拟文件变动后的热更新），载入另一组 mocks
    // 3. 验证旧的 mocks 已被移除，只有新的 mocks 存在

    // 第一次加载
    const initialDir = path.resolve(__dirname, 'mocks/loader-single');
    await loadMocks(initialDir);

    // 验证第一次加载成功
    expect(mockStore.has('/api/test/user')).toBe(true);
    expect(mockStore.size).toBe(2);

    // 第二次加载（模拟热更新）
    const reloadDir = path.resolve(__dirname, 'mocks/loader-reload');
    await loadMocks(reloadDir);

    // 验证结果是否符合预期
    expect(mockStore.size).toBe(1); // 新目录只有一个 mock，所以 size 应该是 1
    expect(mockStore.has('/api/test/status')).toBe(true); // 新的 mock 存在
    expect(mockStore.has('/api/test/user')).toBe(false); // 旧的 mock 已经被清除
    expect(mockStore.has('/api/test/settings')).toBe(false); // 旧的 mock 已经被清除
  });

  // 快照测试（可选，但推荐）
  // 快照测试可以轻松地验证复杂对象的结构是否符合预期。
  it('should match the snapshot of a loaded mock item', async () => {
    const singleFileDir = path.resolve(__dirname, 'mocks/loader-single');
    await loadMocks(singleFileDir);

    const userMock = mockStore.get('/api/test/user');
    // Vitest 会自动在 __snapshots__ 目录下创建一个快照文件。
    // 如果 userMock 的结构发生变化，测试将会失败，提醒你检查变更。
    expect(userMock).toMatchSnapshot();
  });
});
