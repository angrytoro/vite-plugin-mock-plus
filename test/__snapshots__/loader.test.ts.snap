// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mock Loader > should match the snapshot of a loaded mock item 1`] = `
{
  "method": "GET",
  "response": [Function],
}
`;
